File: MarketFlutter/lib/models/order.dart
import 'dart:convert';

import 'package:market/models/offer.dart';
import 'package:market/models/offer_type.dart';


class Order{
  int id;
  double quantity;
  double price;
  String? address;
  int offerId;
  String buyerId;
  String sellerId;
  String status;
  DateTime? dateOrdered;
  DateTime? dateDelivered;
  String title;
  int? billingDetailsId;
  int offerTypeId;
  OfferType? offerType;
  Offer? offer;

  Order({ this.id=0,  this.quantity=0, this.price=0,
     this.address, required this.offerId,
    required this.buyerId, required this.sellerId, required this.offerTypeId, this.dateOrdered, this.title = "none", required this.status, this.dateDelivered,this.billingDetailsId, this.offer, this.offerType});

  factory Order.fromJson(Map<String, dynamic> json) {
    print("Parsing order");

    print(json);
    print("price");
    print(double.parse(json['price'].toString()));

    Order res = Order(
      id: json['id'] as int,
      quantity: json['quantity']+.0 as double,
      price: json['price']+.0 as double,
      address: json['address'] as String,
      offerId: json['offerId'] as int,
      buyerId: json['buyerId'] as String,
      offer: json['offer'] != null ? Offer.fromJson(json['offer']) : null,
      billingDetailsId: json['billingDetailsId'] as int?,
      status: json['status'] as String,
      sellerId: json['sellerId'] as String,
      dateOrdered: DateTime.parse(json['dateOrdered']),
      dateDelivered: json['dateDelivered'] != null ? DateTime.parse(json['dateDelivered']) : null,
      title: json['title'] as String,
      offerTypeId: json['offerType']['id'] as int,
      offerType: OfferType.fromJson(json['offerType']),
    );

    print("DONE");

    return res;
  }

  factory Order.fromStorageJson(Map<String, dynamic> json) {
    print(json);
    Order res = Order(
      id: json['id'] as int,
      quantity: json['quantity']+.0 as double,
      price: json['price']+.0 as double,
      offerId: json['offerId'] as int,
      buyerId: json['buyerId'] as String,
      sellerId: json['sellerId'] as String,
      title: json['title'] as String,
      offer: Offer.fromJson(json['offer']),
      offerTypeId: json['offerType']['id'] as int,
      offerType: OfferType.fromJson(json['offerType']),
      status: json['status'] as String,
    );
    print("done");
    return res;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'quantity': quantity,
      'price': price,
      'address': address,
      'offerId': offerId,
      'buyerId': buyerId,
      'sellerId': sellerId,
      'title': title,
      'offer': offer?.toJson(),
      'offerType': offerType?.toJson(),
      'offerTypeId': offerTypeId,
      'status': status,
      //'offerType': offerType,
    };
  }
}

File: MarketPortal/Market/Data/Models/Order.cs
﻿using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using NuGet.Protocol.Plugins;
using Market.Data.Models;

namespace Market.Data.Models
{

    public class Order
    {
        [Key]
        public int Id { get; set; }
        [Required]
        public string Title { get; set; }

        [Required]
        public string Status { get; set; }


        [Required]
        public decimal Quantity { get; set; }
        [Required]
        public decimal Price { get; set; }
        [Required]
        public string Address { get; set; }

        [Required]
        [ForeignKey(nameof(Offer))]
        public int OfferId { get; set; }
        public Offer? Offer { get; set; }

        public Guid BuyerId { get; set; }
        public User? Buyer { get; set; }

        public Guid? SellerId { get; set; }
        public User? Seller { get; set; }

        [Required]
        public DateTime DateOrdered { get; set; }

        public DateTime? DateDelivered { get; set; } = null;

        [Required]
        [ForeignKey(nameof(OfferType))]
        public int OfferTypeId { get; set; }

        public OfferType? OfferType { get; set; }

        [Required]
        [ForeignKey(nameof(BillingDetails))]

        public int BillingDetailsId { get; set; }
        public BillingDetails? BillingDetails { get; set; }
    }
}
